File: menubar/sidecar/slam_lm_bridge/protocol.py
"""Wire protocol for the SlamLM bridge.

This module is the Python mirror of ``menubar/PROTOCOL.md``: it owns the
encode/decode helpers, the thread-safe stdout writer and the dataclass layer
for every type that crosses the wire. Field names are camelCase on purpose —
they *are* the wire keys, so nothing has to be renamed on the way out.

Nothing here measures anything; metrics live in ``metrics.py``.
"""

from __future__ import annotations

import json
import sys
import threading
import time
from dataclasses import asdict, dataclass
from typing import Any, Dict, List, Optional

PROTOCOL_VERSION = 1
BRIDGE_NAME = "slam-lm-bridge"

#: Canonical display order for categories. Used for `Model.categories` and,
#: filtered to what was discovered, for `hello.categories`.
CATEGORY_ORDER = (
    "Chat",
    "Code",
    "Vision",
    "Embedding",
    "Audio",
    "Instruct",
    "Multilingual",
    "Reasoning",
)


# MARK: - Encode / decode


def encode(payload: Dict[str, Any]) -> str:
    """Encode one wire line: compact JSON, UTF-8, no embedded newlines.

    ``allow_nan=False`` keeps non-JSON floats (NaN/Inf) out of the protocol;
    every producer of numbers must sanitise them rather than leak them here.
    """
    return json.dumps(
        payload, ensure_ascii=False, separators=(",", ":"), allow_nan=False
    )


def decode(line: str) -> Dict[str, Any]:
    """Decode one wire line into a dict. Raises ``ValueError`` on bad input."""
    parsed = json.loads(line)
    if not isinstance(parsed, dict):
        raise ValueError("wire line is not a JSON object")
    return parsed


def reply(id: int, result: Dict[str, Any]) -> Dict[str, Any]:
    return {"id": id, "ok": True, "result": result}


def error_reply(id: int, message: str) -> Dict[str, Any]:
    return {"id": id, "ok": False, "error": message}


def event(name: str, data: Dict[str, Any]) -> Dict[str, Any]:
    return {"event": name, "data": data}


class LineWriter:
    """Serialises stdout (and stderr logging) behind one lock.

    Concurrent writer threads — the sampler, the runner worker, the command
    dispatcher and every HTTP handler — therefore never interleave a line.
    """

    def __init__(self, stream: Any = None, err: Any = None):
        self._stream = stream if stream is not None else sys.stdout
        self._err = err if err is not None else sys.stderr
        self._lock = threading.Lock()

    # -- raw send

    def send(self, payload: Dict[str, Any]) -> None:
        try:
            line = encode(payload)
        except (TypeError, ValueError) as exc:  # a producer bug, not a wire case
            self._log("error", f"could not encode {payload!r}: {exc}")
            rid = payload.get("id")
            if isinstance(rid, int) and "result" in payload:
                line = encode(error_reply(rid, f"encode failure: {exc}"))
            else:
                return
        with self._lock:
            try:
                self._stream.write(line + "\n")
                self._stream.flush()
            except OSError as exc:
                # stdout is gone (the app exited); nothing can be reported.
                self._err.write(f"[{BRIDGE_NAME}] error: stdout write failed: {exc}\n")
                self._err.flush()

    # -- envelopes

    def send_reply(self, id: int, result: Dict[str, Any]) -> None:
        self.send(reply(id, result))

    def send_error(self, id: int, message: str) -> None:
        self.send(error_reply(id, message))

    def send_event(self, name: str, data: Dict[str, Any]) -> None:
        self.send(event(name, data))

    # -- logging (stderr is never parsed; the `log` event feeds the app's strip)

    def _log(self, level: str, message: str) -> None:
        with self._lock:
            self._err.write(f"[{BRIDGE_NAME}] {level}: {message}\n")
            self._err.flush()

    def log(self, level: str, message: str, surface: bool = True) -> None:
        self._log(level, message)
        if surface:
            self.send_event("log", LogPayload(level=level, message=message).to_wire())


# MARK: - Wire types


class _Wire:
    """Mixin giving every wire dataclass its exact-key dict form."""

    def to_wire(self) -> Dict[str, Any]:
        return asdict(self)  # type: ignore[arg-type]


@dataclass
class Model(_Wire):
    id: str
    name: str
    params: str
    quant: str
    bytes: int
    path: str
    categories: List[str]
    architecture: str
    contextLength: int
    hasChatTemplate: bool
    lastUsed: float


@dataclass
class Hardware(_Wire):
    model: str
    chip: str
    gpuCores: int
    totalBytes: int
    recommendedBytes: int
    memoryActiveBytes: int
    memoryPeakBytes: int
    cacheBytes: int
    processRssBytes: int
    systemUsedBytes: int
    systemAppBytes: int
    systemWiredBytes: int
    systemCompressedBytes: int
    systemCachedBytes: int
    systemSwapBytes: int


@dataclass
class LiveMetrics(_Wire):
    ts: float
    memoryActiveBytes: int
    memoryPeakBytes: int
    memoryCacheBytes: int
    memoryTotalBytes: int
    memoryRecommendedBytes: int
    processRssBytes: int
    systemUsedBytes: int
    systemAppBytes: int
    systemWiredBytes: int
    systemCompressedBytes: int
    systemCachedBytes: int
    systemSwapBytes: int
    decodeTps: float
    prefillTps: float
    ttftMs: float
    tokensGenerated: int
    requests: int
    status: str
    phase: str
    model: Optional[str]
    loadMs: float


@dataclass
class RequestRecord(_Wire):
    request: int
    model: str
    promptTokens: int
    genTokens: int
    ttftMs: float
    prefillTps: float
    decodeTps: float
    peakMemBytes: int
    startedAt: float
    totalMs: float
    finishReason: str


@dataclass
class StatePayload(_Wire):
    status: str
    phase: str
    model: Optional[str]
    message: Optional[str] = None


@dataclass
class TokenPayload(_Wire):
    request: int
    index: int
    text: str
    ttsMs: float


@dataclass
class LogPayload(_Wire):
    level: str
    message: str


#: Requests arriving over HTTP get ids from this offset so they can never be
#: confused with ids chosen by the app for its own `generate` commands.
HTTP_REQUEST_BASE = 1_000_000


def sanitise_float(value: Any) -> float:
    """Clamp NaN/Inf to 0.0 so `encode` can never see a non-JSON float."""
    try:
        out = float(value)
    except (TypeError, ValueError):
        return 0.0
    if out != out or out in (float("inf"), float("-inf")):
        return 0.0
    return out


def monotonic_ms(since: float) -> float:
    return (time.perf_counter() - since) * 1000.0

File: menubar/Sources/SlamLM/Protocol.swift
import Combine
import Foundation

// MARK: - Wire types
//
// Mirrors `menubar/PROTOCOL.md` exactly; both sides must agree on these keys.
// Wire structs stay dumb (no logic, no optionals unless the protocol says a
// field is nullable) so a decode failure is always a real contract mismatch.

/// One locally available model. `id` is the repo id and the unique key.
struct ModelInfo: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let params: String
    let quant: String
    let bytes: Int
    let path: String
    let categories: [String]
    let architecture: String
    let contextLength: Int
    let hasChatTemplate: Bool
    let lastUsed: Double
}

/// Machine facts plus a live memory sample. Sourced in the bridge from
/// `sysctl`, `system_profiler`, Mach `host_statistics64` and `mx.device_info()`;
/// never estimated.
struct HardwareInfo: Codable, Hashable {
    let model: String
    let chip: String
    let gpuCores: Int
    let totalBytes: Int
    let recommendedBytes: Int
    let memoryActiveBytes: Int
    let memoryPeakBytes: Int
    let cacheBytes: Int
    let processRssBytes: Int
    let systemUsedBytes: Int
    let systemAppBytes: Int
    let systemWiredBytes: Int
    let systemCompressedBytes: Int
    let systemCachedBytes: Int
    let systemSwapBytes: Int
}

/// Runner status as it appears on the wire.
enum RunnerStatus: String, Codable, CaseIterable {
    case idle, loading, ready, generating, error
}

/// What the runner is spending time on right now.
enum RunnerPhase: String, Codable, CaseIterable {
    case idle, load, prefill, decode
}

/// One 5 Hz telemetry sample. Every field is a real measurement; a field with
/// no data yet is `0` and the UI must render an empty state, not a placeholder.
struct LiveMetrics: Codable, Identifiable, Hashable {
    let ts: Double
    let memoryActiveBytes: Int
    let memoryPeakBytes: Int
    let memoryCacheBytes: Int
    let memoryTotalBytes: Int
    let memoryRecommendedBytes: Int
    let processRssBytes: Int
    /// Machine-wide, Activity Monitor's definitions. Not this process's share.
    let systemUsedBytes: Int
    let systemAppBytes: Int
    let systemWiredBytes: Int
    let systemCompressedBytes: Int
    let systemCachedBytes: Int
    let systemSwapBytes: Int
    let decodeTps: Double
    let prefillTps: Double
    let ttftMs: Double
    let tokensGenerated: Int
    let requests: Int
    let status: String
    let phase: String
    let model: String?
    let loadMs: Double

    var id: Double { ts }
    var runnerStatus: RunnerStatus { RunnerStatus(rawValue: status) ?? .idle }
    var runnerPhase: RunnerPhase { RunnerPhase(rawValue: phase) ?? .idle }
}

/// One completed (or cancelled, or failed) request.
struct RequestRecord: Codable, Identifiable, Hashable {
    let request: Int
    let model: String
    let promptTokens: Int
    let genTokens: Int
    let ttftMs: Double
    let prefillTps: Double
    let decodeTps: Double
    let peakMemBytes: Int
    let startedAt: Double
    let totalMs: Double
    let toolCalls: Int
    let finishReason: String

    var id: Int { request }
    var failed: Bool { finishReason == "error" }
}

// MARK: - Result payloads

struct HelloResult: Codable {
    let protocolVersion: Int
    let bridge: String
    let hardware: HardwareInfo
    let categories: [String]

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case bridge, hardware, categories
    }
}

struct CatalogResult: Codable { let models: [ModelInfo] }
struct LoadResult: Codable { let model: String; let loadMs: Double; let memoryBytes: Int }
struct CancelResult: Codable { let cancelled: Bool }
struct ServeResult: Codable { let port: Int; let url: String }
struct StopServeResult: Codable { let stopped: Bool }
struct EmptyPayload: Codable {}

// MARK: - Event payloads

struct StatePayload: Codable {
    let status: String
    let phase: String
    let model: String?
    let message: String?
}

struct TokenPayload: Codable {
    let request: Int
    let index: Int
    let text: String
    let ttsMs: Double
}

struct LogPayload: Codable {
    let level: String
    let message: String
}

/// One tool call or its result. A call carries the arguments the model asked
/// with; a result carries what came back. `ok` is false for a refusal or a
/// failure, which the model is shown rather than swallowed.
struct ToolEvent: Codable, Identifiable, Hashable {
    let request: Int
    let round: Int
    let phase: String
    let name: String
    let arguments: ToolArguments?
    let ok: Bool
    let summary: String
    let detail: String?

    var id: String { "\(request)-\(round)-\(phase)-\(name)" }
    var isCall: Bool { phase == "call" }
    var failed: Bool { !ok }
}

/// Tool arguments arrive as an arbitrary JSON object. The panel only ever shows
/// them, so each value is flattened to display text here rather than pulling a
/// JSON value type through the whole app.
struct ToolArguments: Codable, Hashable {
    let pairs: [(String, String)]

    /// Only decoding constructs these in the app; this exists so tests and
    /// offscreen harnesses can build one.
    init(pairs: [(String, String)]) { self.pairs = pairs }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DynamicKey.self)
        pairs = try container.allKeys
            .map { ($0.stringValue, try container.decode(AnyJSONText.self, forKey: $0).text) }
            .sorted { $0.0 < $1.0 }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: DynamicKey.self)
        for (key, value) in pairs {
            try container.encode(value, forKey: DynamicKey(stringValue: key)!)
        }
    }

    /// `key: value, key: value`, for a one-line caption.
    var summary: String { pairs.map { "\($0.0): \($0.1)" }.joined(separator: ", ") }

    static func == (lhs: ToolArguments, rhs: ToolArguments) -> Bool {
        lhs.pairs.elementsEqual(rhs.pairs, by: { $0.0 == $1.0 && $0.1 == $1.1 })
    }

    func hash(into hasher: inout Hasher) {
        for pair in pairs { hasher.combine(pair.0); hasher.combine(pair.1) }
    }
}

/// Display text for any JSON value: strings as themselves, everything else in
/// compact JSON form. Recursive, so nested objects and arrays survive.
struct AnyJSONText: Decodable {
    let text: String

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode(String.self) { text = value; return }
        if let value = try? container.decode(Bool.self) { text = value ? "true" : "false"; return }
        if let value = try? container.decode(Int.self) { text = String(value); return }
        if let value = try? container.decode(Double.self) { text = String(value); return }
        if container.decodeNil() { text = "null"; return }
        if let values = try? container.decode([AnyJSONText].self) {
            text = "[" + values.map(\.text).joined(separator: ", ") + "]"
            return
        }
        let keyed = try decoder.container(keyedBy: DynamicKey.self)
        let parts = try keyed.allKeys.map { key -> String in
            let value = try keyed.decode(AnyJSONText.self, forKey: key)
            return "\(key.stringValue): \(value.text)"
        }
        text = "{" + parts.joined(separator: ", ") + "}"
    }
}

struct DynamicKey: CodingKey {
    let stringValue: String
    var intValue: Int? { nil }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { nil }
}

/// Reply envelope: `{"id":Int,"ok":Bool,"result":{…}}` or `{"id":Int,"ok":Bool,"error":String}`.
struct WireReply<Payload: Decodable>: Decodable {
    let id: Int
    let ok: Bool
    let error: String?
    let result: Payload?
}

typealias AckReply = WireReply<EmptyPayload>

struct WireEvent<P: Decodable>: Decodable {
    let event: String
    let data: P
}

/// Probe used to route a stdout line before decoding its payload.
struct WireProbe: Decodable {
    let id: Int?
    let ok: Bool?
    let error: String?
    let event: String?
}

// MARK: - The one state object both surfaces read
//
// State lives here, transport does not: `BridgeClient` (picker surface) decodes
// process output and pushes it in, and every view observes this single object.
// The analytics surface therefore depends only on this file, never on the
// transport.

final class MetricsBus: ObservableObject {
    /// Ring buffers. Sized for the panel's own history charts.
    static let historyLimit = 600
    static let requestLimit = 100
    static let toolEventLimit = 200

    @Published var status: RunnerStatus = .idle
    @Published var phase: RunnerPhase = .idle
    @Published var models: [ModelInfo] = []
    @Published var selected: ModelInfo?
    @Published var hardware: HardwareInfo?
    @Published var categories: [String] = []
    @Published var live: LiveMetrics?
    @Published var history: [LiveMetrics] = []
    @Published var requests: [RequestRecord] = []
    @Published var servingURL: String?
    @Published var logLine: String?
    @Published var busy = false
    @Published var errorText: String?
    @Published var bridgeReady = false
    /// Text streamed for the most recent app-initiated request, newest last.
    @Published var streamText = ""
    /// Tool calls and results for the most recent request, in order.
    @Published var toolEvents: [ToolEvent] = []
    /// Whether the model may call the read-only tools. Read by the store when it
    /// sends a prompt, so both surfaces agree on the setting.
    @Published var toolsEnabled = true

    /// Append a telemetry sample, keeping the buffer bounded.
    func ingest(_ sample: LiveMetrics) {
        live = sample
        status = sample.runnerStatus
        phase = sample.runnerPhase
        history.append(sample)
        if history.count > Self.historyLimit {
            history.removeFirst(history.count - Self.historyLimit)
        }
    }

    /// Record a finished request, newest first.
    func ingest(_ record: RequestRecord) {
        requests.insert(record, at: 0)
        if requests.count > Self.requestLimit {
            requests.removeLast(requests.count - Self.requestLimit)
        }
    }

    func apply(_ state: StatePayload) {
        status = RunnerStatus(rawValue: state.status) ?? status
        phase = RunnerPhase(rawValue: state.phase) ?? phase
    }

    /// Record a tool call or result, keeping the trace bounded.
    func ingest(_ event: ToolEvent) {
        toolEvents.append(event)
        if toolEvents.count > Self.toolEventLimit {
            toolEvents.removeFirst(toolEvents.count - Self.toolEventLimit)
        }
    }

    func clearStream() {
        streamText = ""
        toolEvents = []
    }
}

/// Panel tabs. `models` is the picker, `analytics` the expanded metrics surface.
enum PanelTab: String, CaseIterable, Identifiable {
    case models, analytics
    var id: String { rawValue }
    var title: String { self == .models ? "Models" : "Analytics" }
    var symbol: String { self == .models ? "square.stack.3d.up" : "chart.xyaxis.line" }
}

/// Commands the analytics surface may invoke, injected by the picker surface so
/// the panel never depends on the transport or the store.
struct AnalyticsActions {
    var send: (String) -> Void
    var cancel: () -> Void
    /// Return to the model picker. The panel owns its tab state, so it supplies
    /// this; the analytics surface must never be a dead end.
    var showModels: () -> Void
}

// MARK: - Runtime configuration

/// Where the bridge lives and what it is launched with. Resolved once at
/// startup from the command line and environment; see `PROTOCOL.md`.
struct RunnerConfig {
    var pythonPath: String
    var bridgeDir: String
    /// Repo checkout holding `mlx_lm` and `.venv`. The app passes this to the
    /// bridge so it never has to guess where the package lives.
    var repoRoot: String
    var port: Int
    var tab: PanelTab
    var preview: Bool
    var snapshotPath: String?
    var snapshotAnalytics: String?
    var model: String?
    /// A token budget for a snapshot run; absent means no budget.
    var tokens: Int?
    /// Overrides the built-in snapshot prompt, so the capture can exercise a
    /// particular shape of output (markdown, code, a long list).
    var prompt: String?
    /// How many real runs `--snapshot-analytics` drives before rendering.
    var requests: Int
    /// Loads the model before a plain `--snapshot` too, so the captured panel
    /// shows the running state rather than an idle one.
    var loadForSnapshot: Bool

    static func resolve() -> RunnerConfig {
        var args = Array(CommandLine.arguments.dropFirst())
        func takeFlag(_ name: String) -> String? {
            guard let i = args.firstIndex(of: name) else { return nil }
            guard i + 1 < args.count else { return "" }
            let value = args[i + 1]
            args.removeSubrange(i...(i + 1))
            return value
        }
        let env = ProcessInfo.processInfo.environment
        let snapshot = takeFlag("--snapshot")
        let snapshotAnalytics = takeFlag("--snapshot-analytics")
        let tab = takeFlag("--tab").flatMap(PanelTab.init(rawValue:)) ?? .models
        let port = takeFlag("--port").flatMap(Int.init) ?? 8712
        let model = RunnerConfig.nonEmpty(takeFlag("--model"))
        let tokens = takeFlag("--tokens").flatMap(Int.init)
        let requests = max(1, takeFlag("--requests").flatMap(Int.init) ?? 1)
        let prompt = RunnerConfig.nonEmpty(takeFlag("--prompt"))
        let loadForSnapshot = args.contains("--load")
        let preview = args.contains("--preview") || snapshot != nil || snapshotAnalytics != nil

        let bridgeDir: String
        if let explicit = env["SLAM_LM_BRIDGE_DIR"], !explicit.isEmpty {
            bridgeDir = explicit
        } else if let res = Bundle.main.resourcePath,
                  FileManager.default.fileExists(atPath: res + "/sidecar/slam_lm_bridge") {
            bridgeDir = res + "/sidecar"
        } else {
            // Running straight out of `.build/debug` during development.
            let here = URL(fileURLWithPath: CommandLine.arguments[0]).resolvingSymlinksInPath()
            bridgeDir = here.deletingLastPathComponent().path
        }

        // The bundle records the interpreter and checkout it was built against
        // (`build.sh` writes it), so a Finder launch never has to guess them.
        let runtime = RunnerConfig.bundledRuntime()

        let repoRoot: String
        if let explicit = env["SLAM_LM_REPO"], !explicit.isEmpty {
            repoRoot = explicit
        } else if let recorded = runtime.repo, RunnerConfig.looksLikeRepo(recorded) {
            repoRoot = recorded
        } else {
            repoRoot = RunnerConfig.findRepoRoot(bridgeDir: bridgeDir)
        }

        let pythonPath: String
        if let explicit = env["SLAM_LM_PYTHON"], !explicit.isEmpty {
            pythonPath = explicit
        } else if let recorded = runtime.python,
                  FileManager.default.isExecutableFile(atPath: recorded) {
            pythonPath = recorded
        } else {
            pythonPath = repoRoot + "/.venv/bin/python"
        }

        return RunnerConfig(
            pythonPath: pythonPath,
            bridgeDir: bridgeDir,
            repoRoot: repoRoot,
            port: port,
            tab: tab,
            preview: preview,
            snapshotPath: RunnerConfig.nonEmpty(snapshot),
            snapshotAnalytics: RunnerConfig.nonEmpty(snapshotAnalytics),
            model: model,
            tokens: tokens,
            prompt: prompt,
            requests: requests,
            loadForSnapshot: loadForSnapshot
        )
    }

    private struct BundledRuntime: Decodable {
        let python: String?
        let repo: String?
    }

    private static func bundledRuntime() -> BundledRuntime {
        let empty = BundledRuntime(python: nil, repo: nil)
        guard let resources = Bundle.main.resourcePath,
              let data = FileManager.default.contents(atPath: resources + "/runtime.json"),
              let parsed = try? JSONDecoder().decode(BundledRuntime.self, from: data)
        else { return empty }
        return parsed
    }

    /// A directory is the checkout when it holds `mlx_lm`, and in a development
    /// tree the virtualenv beside it.
    private static func looksLikeRepo(_ path: String) -> Bool {
        let fm = FileManager.default
        return fm.fileExists(atPath: path + "/mlx_lm/__init__.py")
            || fm.fileExists(atPath: path + "/.venv/bin/python")
    }

    /// Empty and absent flags mean the same thing to the caller.
    private static func nonEmpty(_ value: String?) -> String? {
        guard let value, !value.isEmpty else { return nil }
        return value
    }

    /// Walk up from the bridge directory looking for the checkout: the bridge
    /// sits at `menubar/sidecar` in a development tree and at
    /// `Contents/Resources/sidecar` inside the bundle.
    private static func findRepoRoot(bridgeDir: String) -> String {
        var dir = URL(fileURLWithPath: bridgeDir).standardizedFileURL
        for _ in 0..<9 {
            if looksLikeRepo(dir.path) { return dir.path }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }
        return FileManager.default.currentDirectoryPath
    }
}
